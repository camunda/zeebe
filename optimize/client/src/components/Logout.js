/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {useEffect} from 'react';
import {withRouter} from 'react-router-dom';
import {get} from 'request';
import {withErrorHandling} from 'HOC';
import {addNotification} from 'notifications';
import {t} from 'translation';

export function Logout({mightFail, history}) {
  useEffect(() => {
    (async () => {
      await mightFail(
        get('api/authentication/logout'),
        () => addNotification({text: t('navigation.logoutSuccess')}),
        () => addNotification({text: t('navigation.logoutFailed'), type: 'error'})
      );
      history.replace('/');
    })();
  }, [mightFail, history]);

  return null;
}

export default withRouter(withErrorHandling(Logout));
