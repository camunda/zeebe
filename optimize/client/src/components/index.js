/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

export {Analysis} from './Analysis';
export {Footer} from './Footer';
export {Header} from './Header';
export {PrivateRoute} from './PrivateRoute';
export {Dashboard} from './Dashboards';
export {Report} from './Reports';
export {Sharing} from './Sharing';
export {Home, Collection} from './Home';
export {default as Logout} from './Logout';
export {Processes, ProcessReport} from './Processes';
