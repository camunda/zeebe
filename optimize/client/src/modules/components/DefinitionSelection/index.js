/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

export {default as DefinitionSelection} from './DefinitionSelection';
export {default as VersionPopover} from './VersionPopover';
export {default as TenantPopover} from './TenantPopover';
