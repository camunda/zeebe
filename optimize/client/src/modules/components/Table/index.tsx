/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

export {default as Table} from './Table';

export type {Body as TableBody, Head as TableHead} from './Table';
