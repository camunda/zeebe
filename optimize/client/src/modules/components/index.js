/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

export {Table} from './Table';
export {Button} from './Button';
export {Modal} from './Modal';
export {Dropdown} from './Dropdown';
export {Deleter} from './Deleter';
export {Input} from './Input';
export {Select} from './Select';
export {BPMNDiagram, ClickBehavior, PartHighlight} from './BPMNDiagram';
export {DMNDiagram} from './DMNDiagram';
export {CopyToClipboard} from './CopyToClipboard';
export {ReportRenderer, NoDataNotice} from './ReportRenderer';
export {ErrorBoundary} from './ErrorBoundary';
export {Icon} from './Icon';
export {ActionItem} from './ActionItem';
export {ShareEntity} from './ShareEntity';
export {DashboardRenderer, DiagramScrollLock} from './DashboardRenderer';
export {DefinitionSelection, VersionPopover, TenantPopover} from './DefinitionSelection';
export {TargetValueBadge} from './TargetValueBadge';
export {ErrorPage} from './ErrorPage';
export {Labeled} from './Labeled';
export {LabeledInput} from './LabeledInput';
export {DatePicker} from './DatePicker';
export {LoadingIndicator} from './LoadingIndicator';
export {ColorPicker} from './ColorPicker';
export {EntityName} from './EntityName';
export {EntityNameForm} from './EntityNameForm';
export {SubNav, NavItem} from './navigation';
export {HeatmapOverlay} from './HeatmapOverlay';
export {EntityList} from './EntityList';
export {UserTypeahead} from './UserTypeahead';
export {DateRangeInput} from './DateRangeInput';
export {LastModifiedInfo} from './LastModifiedInfo';
export {ReportDetails} from './ReportDetails';
export {InstanceCount} from './InstanceCount';
export {DocsLink} from './DocsLink';
export {Tabs} from './Tabs';
export {Checklist} from './Checklist';
export {Tooltip} from './Tooltip';
export {SelectionPreview} from './SelectionPreview';
export {DownloadButton} from './DownloadButton';
export {BulkDeleter} from './BulkDeleter';
export {AlertModal} from './AlertModal';
export {AlertsDropdown} from './AlertsDropdown';
export {DurationChart} from './DurationChart';
export {ReportTemplateModal, DashboardTemplateModal} from './TemplateModal';
export {KpiCreationModal} from './KpiCreationModal';
export {EmptyState} from './EmptyState';
export {PageTitle} from './PageTitle';
export {TextEditor} from './TextEditor';
export {EntityDescription} from './EntityDescription';
export {Popover} from './Popover';
export {InstanceViewTable} from './InstanceViewTable';
export {TenantInfo} from './TenantInfo';
export {TargetSelection} from './TargetSelection';
export {Loading} from './Loading';
