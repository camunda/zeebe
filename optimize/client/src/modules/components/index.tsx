/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

export {Button} from './Button';
export {Tabs} from './Tabs';
export {Icon} from './Icon';
export {ActionItem} from './ActionItem';
export {Input} from './Input';
export {LabeledInput} from './LabeledInput';
export {Labeled} from './Labeled';
export {Tooltip} from './Tooltip';
export {Modal} from './Modal';
export {Dropdown} from './Dropdown';
export {Select} from './Select';
export {TextEditor} from './TextEditor';
export {PageTitle} from './PageTitle';
export {EntityDescription} from './EntityDescription';
export {EntityNameForm} from './EntityNameForm';
export {LoadingIndicator} from './LoadingIndicator';
export {DocsLink} from './DocsLink';
export {UserTypeahead} from './UserTypeahead';
export {BPMNDiagram, ClickBehavior} from './BPMNDiagram';
export {DMNDiagram} from './DMNDiagram';
export {DownloadButton} from './DownloadButton';
export {NoDataNotice} from './ReportRenderer';
export {Table} from './Table';
export {TargetValueBadge} from './TargetValueBadge';
export {DatePicker} from './DatePicker';
export {DateRangeInput} from './DateRangeInput';
export {Checklist} from './Checklist';
export {DurationChart} from './DurationChart';
export {ColorPicker} from './ColorPicker';
export {Popover} from './Popover';
export {TenantInfo} from './TenantInfo';
export {VersionPopover, TenantPopover} from './DefinitionSelection';
export {CopyToClipboard} from './CopyToClipboard';
export {ShareEntity} from './ShareEntity';
export {ErrorPage} from './ErrorPage';
export {ErrorBoundary} from './ErrorBoundary';
export {TargetSelection} from './TargetSelection';
// @ts-ignore
export {DefinitionSelection} from './DefinitionSelection';
// @ts-ignore
export {ReportRenderer} from './ReportRenderer';
export {Loading} from './Loading';
export {EntityList} from './EntityList';
export {EmptyState} from './EmptyState';

export type {InputProps} from './Input';
export type {LabeledProps} from './Labeled';
export type {User} from './UserTypeahead';
export type {Canvas, ModdleElement, RegistryElement} from './BPMNDiagram';
export type {TooltipProps} from './Tooltip';
export type {TableBody, TableHead} from './Table';
