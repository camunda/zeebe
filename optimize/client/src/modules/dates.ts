/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {format as formatFns, Locale} from 'date-fns';

export let globalLocale: Locale;

// by providing a default string of 'PP' or any of its variants for `formatStr`
// it will format dates in whichever way is appropriate to the locale
export function format(date: Date | number, formatStr = 'PP'): string {
  return formatFns(date, formatStr, {
    locale: globalLocale,
  });
}

export async function loadDateTranslation(localeCode: string): Promise<void> {
  switch (localeCode) {
    case 'de':
      globalLocale = (await import('date-fns/locale/de')).default;
      break;
    case 'fr':
      globalLocale = (await import('date-fns/locale/fr')).default;
      break;
    case 'es':
      globalLocale = (await import('date-fns/locale/es')).default;
      break;
    case 'hi':
      globalLocale = (await import('date-fns/locale/hi')).default;
      break;
    case 'it':
      globalLocale = (await import('date-fns/locale/it')).default;
      break;
    case 'ja':
      globalLocale = (await import('date-fns/locale/ja')).default;
      break;
    case 'pt':
      globalLocale = (await import('date-fns/locale/pt')).default;
      break;
    case 'ru':
      globalLocale = (await import('date-fns/locale/ru')).default;
      break;
    default:
      break;
  }
}

export const BACKEND_DATE_FORMAT = "yyyy-MM-dd'T'HH:mm:ss.SSSxx";
