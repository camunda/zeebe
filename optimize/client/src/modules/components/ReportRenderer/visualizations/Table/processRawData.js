/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {parseISO} from 'date-fns';

import {format} from 'dates';
import {formatters} from 'services';
import {t} from 'translation';
import {Button} from 'components';

import {
  cockpitLink,
  getNoDataMessage,
  isVisibleColumn,
  getLabelWithType,
  sortColumns,
} from './service';

const {duration} = formatters;

const instanceColumns = {
  process: [
    'processDefinitionKey',
    'processDefinitionId',
    'processInstanceId',
    'businessKey',
    'startDate',
    'endDate',
    'duration',
    'engineName',
    'tenantId',
  ],
  decision: [
    'decisionDefinitionKey',
    'decisionDefinitionId',
    'decisionInstanceId',
    'processInstanceId',
    'evaluationDateTime',
    'engineName',
    'tenantId',
  ],
};

export const OBJECT_VARIABLE_IDENTIFIER = '<<OBJECT_VARIABLE_VALUE>>';

export default function processRawData({
  report: {
    reportType,
    data: {
      configuration: {tableColumns},
    },
    result: {data: result},
  },
  camundaEndpoints = {},
  processVariables = [],
  onVariableView,
}) {
  const instanceProps = instanceColumns[reportType].filter((entry) =>
    isVisibleColumn(entry, tableColumns)
  );

  const {
    counts,
    variables,
    flowNodeDurations,
    inputVariables: inputVars,
    outputVariables: outputVars,
  } = result[0] || {};

  const countKeys = getVisibleColumns(counts, tableColumns, 'count');

  const variableNames = getVisibleColumns(variables, tableColumns, 'variable');

  const inputVariables = getVisibleColumns(inputVars, tableColumns, 'input');

  const outputVariables = getVisibleColumns(outputVars, tableColumns, 'output');

  const flowNodeDurationNames = getVisibleColumns(flowNodeDurations || {}, tableColumns, 'dur');

  // If all columns are excluded return a message to enable one
  if (
    instanceProps.length +
      countKeys.length +
      variableNames.length +
      inputVariables.length +
      outputVariables.length +
      flowNodeDurationNames.length ===
    0
  ) {
    return getNoDataMessage();
  }

  const body = result.map((instance) => {
    const row = instanceProps.map((entry) => {
      if (entry === 'processInstanceId') {
        return cockpitLink(camundaEndpoints, instance, 'process');
      }

      if (entry === 'decisionInstanceId') {
        return cockpitLink(camundaEndpoints, instance, 'decision');
      }

      if (
        (entry === 'startDate' || entry === 'endDate' || entry === 'evaluationDateTime') &&
        instance[entry]
      ) {
        return format(parseISO(instance[entry]), "yyyy-MM-dd HH:mm:ss 'UTC'X");
      }

      if (entry === 'duration') {
        return duration(instance[entry]);
      }

      return instance[entry];
    });

    const onVariableClick = (variableName) =>
      onVariableView(variableName, instance.processInstanceId, instance.processDefinitionKey);

    return [
      ...row,
      ...getVariableValues(countKeys, instance.counts),
      ...getVariableValues(variableNames, instance.variables, onVariableClick),
      ...getVariableValues(inputVariables, instance.inputVariables),
      ...getVariableValues(outputVariables, instance.outputVariables),
      ...flowNodeDurationNames.map((key) => duration(instance.flowNodeDurations[key]?.value)),
    ];
  });

  const head = instanceProps
    .map((key) => {
      const label = t('report.table.rawData.' + key);
      return {id: key, label, title: label};
    })
    .concat(
      countKeys.map((key) => {
        const name = t('report.table.rawData.' + key);
        return {
          type: 'counts',
          // we dont give them any prefix to keep compatibility with included/excluded columns
          id: key,
          label: getLabelWithType(name, 'count'),
          title: name,
        };
      })
    )
    .concat(
      variableNames.map((name) => ({
        type: 'variables',
        id: 'variable:' + name,
        label: getLabelWithType(getVariableLabel(processVariables, name) || name, 'variable'),
        title: name,
      }))
    )
    .concat(
      inputVariables.map((key) => {
        const {name, id} = inputVars[key];
        const label = name || id;
        return {
          type: 'inputVariables',
          id: 'inputVariable:' + id,
          label: getLabelWithType(label, 'inputVariable'),
          title: label,
        };
      })
    )
    .concat(
      outputVariables.map((key) => {
        const {name, id} = outputVars[key];
        const label = name || id;
        return {
          type: 'outputVariables',
          id: 'outputVariable:' + id,
          label: getLabelWithType(label, 'outputVariable'),
          title: label,
        };
      })
    )
    .concat(
      flowNodeDurationNames.map((key) => {
        const {name} = flowNodeDurations[key];
        const label = name || key;
        return {
          type: 'flowNodeDurations',
          id: 'flowNodeDuration:' + key,
          label: getLabelWithType(label, 'flowNodeDuration'),
          title: label,
          sortable: false,
        };
      })
    );
  const {sortedHead, sortedBody} = sortColumns(head, body, tableColumns.columnOrder);

  return {head: sortedHead, body: sortedBody};
}

function getVariableValues(variableKeys, variableValues, onVariableClick) {
  return variableKeys.map((entry) => {
    const variableData = variableValues[entry];

    if (variableData === OBJECT_VARIABLE_IDENTIFIER) {
      return (
        <Button
          className="ObjectViewBtn"
          link
          onClick={() => {
            onVariableClick(entry);
          }}
        >
          {t('common.view')}
        </Button>
      );
    }

    // Output variables have multiple values
    if (variableData?.values) {
      return variableData.values.join(', ');
    }

    if (variableData?.value) {
      return variableData.value.toString();
    }

    if (variableData !== undefined && typeof variableData !== 'object') {
      return variableData.toString();
    }

    return '';
  });
}

function getVariableLabel(variables, name) {
  return variables.find((variable) => variable.name === name)?.label;
}

function getVisibleColumns(column = {}, tableColumns, type) {
  return Object.keys(column).filter((entry) => isVisibleColumn(`${type}:${entry}`, tableColumns));
}
