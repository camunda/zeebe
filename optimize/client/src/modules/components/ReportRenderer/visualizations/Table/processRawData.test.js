/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import React from 'react';
import {parseISO} from 'date-fns';

import {format} from 'dates';
import {NoDataNotice} from 'components';

import processRawData, {OBJECT_VARIABLE_IDENTIFIER} from './processRawData';

describe('Process table', () => {
  const data = {
    configuration: {
      tableColumns: {
        includeNewVariables: false,
        includedColumns: [
          'processInstanceId',
          'processDefinitionId',
          'variable:var1',
          'variable:var2',
          'variable:var3',
          'dur:dur1',
          'dur:dur2',
        ],
        excludedColumns: [],
        columnOrder: [],
      },
    },
  };

  const result = {
    data: [
      {
        processInstanceId: 'foo',
        processDefinitionId: 'bar',
        variables: {
          var1: 12,
          var2: null,
          var3: OBJECT_VARIABLE_IDENTIFIER,
        },
        flowNodeDurations: {
          dur1: {name: 'dur1Name', value: null},
          dur2: {name: null, value: 2000},
        },
      },
      {
        processInstanceId: 'xyz',
        processDefinitionId: 'abc',
        variables: {
          var1: null,
          var2: true,
        },
        flowNodeDurations: {
          dur1: {name: 'dur1name', value: 1000},
          dur2: {name: null, value: null},
        },
      },
    ],
  };

  it('should transform data to table compatible format', () => {
    expect(processRawData({report: {reportType: 'process', data, result}})).toMatchSnapshot();
  });

  it('should not include columns that are hidden', () => {
    const data = {
      configuration: {
        tableColumns: {
          includeNewVariables: false,
          includedColumns: ['processInstanceId'],
          excludedColumns: ['processDefinitionId', 'variable:var1', 'variable:var1', 'dur:dur1'],
          columnOrder: [],
        },
      },
    };
    expect(processRawData({report: {reportType: 'process', data, result}})).toEqual({
      body: [['foo'], ['xyz']],
      head: [{id: 'processInstanceId', label: 'Process instance Id', title: 'Process instance Id'}],
    });
  });

  it('should exclude variable columns using the variable prefix', () => {
    const data = {
      configuration: {
        tableColumns: {
          includeNewVariables: false,
          includedColumns: ['processInstanceId', 'processDefinitionId', 'variable:var1'],
          excludedColumns: ['variable:var1'],
          columnOrder: [],
        },
      },
    };
    expect(processRawData({report: {reportType: 'process', data, result}})).toMatchSnapshot();
  });

  it('should make the processInstanceId a link', () => {
    const cell = processRawData({
      report: {
        reportType: 'process',
        result: {data: [{processInstanceId: '123', engineName: '1'}]},
        data: {
          configuration: {
            tableColumns: {includedColumns: ['processInstanceId', 'engineName'], columnOrder: []},
          },
        },
      },
      camundaEndpoints: {1: {endpoint: 'http://camunda.com', engineName: 'a'}},
    }).body[0][0];

    expect(cell.type).toBe('a');
    expect(cell.props.href).toBe('http://camunda.com/app/cockpit/a/#/process-instance/123');
  });

  it('should format start and end dates', () => {
    // using format here to dynamically return date with client timezone
    const startDate = format(parseISO('2019-06-07'), 'yyyy-MM-dd');
    const endDate = format(parseISO('2019-06-09'), 'yyyy-MM-dd');

    const cells = processRawData({
      report: {
        reportType: 'process',
        result: {
          data: [{startDate, endDate}],
        },
        data: {
          configuration: {
            tableColumns: {includedColumns: ['startDate', 'endDate'], columnOrder: []},
          },
        },
      },
    }).body[0];

    const expectedDateFormat = "yyyy-MM-dd HH:mm:ss 'UTC'X";
    expect(cells[0]).toBe(format(parseISO(startDate), expectedDateFormat));
    expect(cells[1]).toBe(format(parseISO(endDate), expectedDateFormat));
  });

  it('should format duration', () => {
    const cells = processRawData({
      report: {
        reportType: 'process',
        result: {
          data: [{duration: 123023423}],
        },
        data: {
          configuration: {tableColumns: {includedColumns: ['duration'], columnOrder: []}},
        },
      },
    }).body[0];

    expect(cells[0]).toBe('1d 10h 10min 23s 423ms');
  });

  it('should not make the processInstanceId a link if no endpoint is specified', () => {
    const cell = processRawData({
      report: {
        reportType: 'process',
        result: {data: [{processInstanceId: '123', engineName: '1'}]},
        data: {
          configuration: {
            tableColumns: {includedColumns: ['processInstanceId', 'engineName'], columnOrder: []},
          },
        },
      },
    }).body[0][0];

    expect(cell).toBe('123');
  });

  it('should invoke the variable view callback function when viewing an object variable', () => {
    const spy = jest.fn();
    const {body} = processRawData({
      report: {
        reportType: 'process',
        data,
        result,
      },
      onVariableView: spy,
    });

    body[0][4].props.onClick();

    const {processInstanceId, processDefinitionKey} = result.data[0];
    expect(spy).toHaveBeenCalledWith('var3', processInstanceId, processDefinitionKey);
  });

  it('should disable sorting for flow node duration columns', () => {
    const {head} = processRawData({report: {reportType: 'process', data, result}});

    const flowNodeDurationColumns = head.filter((column) => column.type === 'flowNodeDurations');

    expect(flowNodeDurationColumns[0].sortable).toBe(false);
  });
});

describe('Decision table', () => {
  const data = {
    configuration: {
      tableColumns: {
        includeNewVariables: true,
        includedColumns: [],
        excludedColumns: [
          'processDefinitionKey',
          'processDefinitionId',
          'processInstanceId',
          'businessKey',
          'startDate',
          'numberOfIncidents',
          'numberOfOpenIncidents',
          'numberOfUserTasks',
          'endDate',
          'duration',
          'engineName',
          'tenantId',
          'variable:var1',
          'variable:var2',
          'dur:dur1',
        ],
        columnOrder: [],
      },
    },
  };

  it('should show no data message when all column are excluded', () => {
    const result = {
      data: [{decisionInstanceId: 'foo', decisionDefinitionId: 'bar'}],
    };

    expect(processRawData({report: {reportType: 'process', data, result}})).toEqual({
      body: [],
      head: [],
      noData: <NoDataNotice type="info">You need to enable at least one table column</NoDataNotice>,
    });
  });

  it('should show default instances column headers for empty results', () => {
    expect(
      processRawData({report: {reportType: 'process', data, result: {data: []}}})
    ).toMatchSnapshot();
  });

  it('should display decision and process instance ids as links for decision tables', () => {
    const data = {
      configuration: {
        tableColumns: {
          includeNewVariables: false,
          includedColumns: ['decisionInstanceId', 'processInstanceId', 'engineName'],
          excludedColumns: [],
          columnOrder: [],
        },
      },
    };

    const cell = processRawData({
      report: {
        reportType: 'decision',
        result: {
          data: [{decisionInstanceId: '123', processInstanceId: '456', engineName: 'a'}],
        },
        data,
      },
      camundaEndpoints: {a: {endpoint: 'http://camunda.com', engineName: 'a'}},
    }).body[0];

    expect(cell[0].type).toBe('a');
    expect(cell[0].props.href).toBe('http://camunda.com/app/cockpit/a/#/decision-instance/123');
    expect(cell[1].props.href).toBe('http://camunda.com/app/cockpit/a/#/process-instance/456');
  });

  it('should return correct table props', () => {
    const data = {
      configuration: {
        tableColumns: {
          includeNewVariables: false,
          includedColumns: [
            'decisionDefinitionId',
            'decisionInstanceId',
            'processInstanceId',
            'engineName',
            'input:var1',
            'input:var2',
            'output:result',
            'dur:dur1',
          ],
          excludedColumns: [],
          columnOrder: [],
        },
      },
    };

    const result = {
      data: [
        {
          decisionInstanceId: 'foo',
          processInstanceId: '456',
          decisionDefinitionId: 'bar',
          engineName: '1',
          inputVariables: {
            var1: {id: 'var1', value: 12, name: 'Var 1'},
            var2: {id: 'var2', value: null, name: 'Var 2'},
          },
          outputVariables: {
            result: {id: 'result', values: [1], name: 'Result'},
          },
          flowNodeDurations: {
            dur1: {name: 'dur1name', value: 1000},
            dur2: {name: null, value: null},
          },
        },
        {
          decisionInstanceId: 'xyz',
          processInstanceId: '456',
          decisionDefinitionId: 'abc',
          engineName: '1',
          inputVariables: {
            var1: {id: 'var1', value: null, name: 'Var 1'},
            var2: {id: 'var2', value: true, name: 'Var 2'},
          },
          outputVariables: {
            result: {id: 'result', values: [8], name: 'Result'},
          },
          flowNodeDurations: {
            dur1: {name: 'dur1Name', value: null},
            dur2: {name: null, value: 2000},
          },
        },
      ],
    };

    expect(processRawData({report: {result, reportType: 'decision', data}})).toMatchSnapshot();
  });

  it('should show no data message when all column are excluded', () => {
    const result = {
      data: [
        {
          decisionInstanceId: 'foo',
          decisionDefinitionId: 'bar',
          inputVariables: {
            var1: {id: 'var1', value: 12, name: 'Var 1'},
          },
          outputVariables: {
            result: {id: 'result', values: [1], name: 'Result'},
          },
          flowNodeDurations: {
            dur1: {name: 'dur1Name', value: null},
          },
        },
      ],
    };
    expect(
      processRawData({
        report: {
          reportType: 'decision',
          result,
          data: {
            configuration: {
              tableColumns: {
                columnOrder: [],
                includeNewVariables: true,
                includedColumns: [],
                excludedColumns: [
                  'decisionDefinitionKey',
                  'decisionDefinitionId',
                  'decisionInstanceId',
                  'processInstanceId',
                  'evaluationDateTime',
                  'engineName',
                  'tenantId',
                  'input:var1',
                  'input:var2',
                  'output:result',
                  'dur:dur1',
                ],
              },
            },
          },
        },
      })
    ).toEqual({
      body: [],
      head: [],
      noData: <NoDataNotice type="info">You need to enable at least one table column</NoDataNotice>,
    });
  });

  it('should only show instance table column if result data is empty', () => {
    expect(
      processRawData({report: {data, reportType: 'decision', result: {data: []}}})
    ).toMatchSnapshot();
  });
});
