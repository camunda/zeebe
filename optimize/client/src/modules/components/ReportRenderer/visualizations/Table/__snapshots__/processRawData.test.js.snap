// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Decision table should only show instance table column if result data is empty 1`] = `
Object {
  "body": Array [],
  "head": Array [
    Object {
      "id": "decisionDefinitionKey",
      "label": "Decision definition key",
      "title": "Decision definition key",
    },
    Object {
      "id": "decisionDefinitionId",
      "label": "Decision definition Id",
      "title": "Decision definition Id",
    },
    Object {
      "id": "decisionInstanceId",
      "label": "Decision instance Id",
      "title": "Decision instance Id",
    },
    Object {
      "id": "evaluationDateTime",
      "label": "Evaluation date",
      "title": "Evaluation date",
    },
  ],
}
`;

exports[`Decision table should return correct table props 1`] = `
Object {
  "body": Array [
    Array [
      "bar",
      "foo",
      "456",
      "1",
      "12",
      "",
      "1",
      "1s",
    ],
    Array [
      "abc",
      "xyz",
      "456",
      "1",
      "",
      "true",
      "8",
      "--",
    ],
  ],
  "head": Array [
    Object {
      "id": "decisionDefinitionId",
      "label": "Decision definition Id",
      "title": "Decision definition Id",
    },
    Object {
      "id": "decisionInstanceId",
      "label": "Decision instance Id",
      "title": "Decision instance Id",
    },
    Object {
      "id": "processInstanceId",
      "label": "Process instance Id",
      "title": "Process instance Id",
    },
    Object {
      "id": "engineName",
      "label": "Engine name",
      "title": "Engine name",
    },
    Object {
      "id": "inputVariable:var1",
      "label": <React.Fragment>
        <span
          className="variableExtension"
        >
          InputVar
          : 
        </span>
        Var 1
      </React.Fragment>,
      "title": "Var 1",
      "type": "inputVariables",
    },
    Object {
      "id": "inputVariable:var2",
      "label": <React.Fragment>
        <span
          className="variableExtension"
        >
          InputVar
          : 
        </span>
        Var 2
      </React.Fragment>,
      "title": "Var 2",
      "type": "inputVariables",
    },
    Object {
      "id": "outputVariable:result",
      "label": <React.Fragment>
        <span
          className="variableExtension"
        >
          OutputVar
          : 
        </span>
        Result
      </React.Fragment>,
      "title": "Result",
      "type": "outputVariables",
    },
    Object {
      "id": "flowNodeDuration:dur1",
      "label": <React.Fragment>
        <span
          className="variableExtension"
        >
          Dur
          : 
        </span>
        dur1name
      </React.Fragment>,
      "sortable": false,
      "title": "dur1name",
      "type": "flowNodeDurations",
    },
  ],
}
`;

exports[`Decision table should show default instances column headers for empty results 1`] = `
Object {
  "body": Array [],
  "head": Array [],
  "noData": <NoDataNotice
    type="info"
  >
    You need to enable at least one table column
  </NoDataNotice>,
}
`;

exports[`Process table should exclude variable columns using the variable prefix 1`] = `
Object {
  "body": Array [
    Array [
      "bar",
      "foo",
      "12",
    ],
    Array [
      "abc",
      "xyz",
      "",
    ],
  ],
  "head": Array [
    Object {
      "id": "processDefinitionId",
      "label": "Process definition Id",
      "title": "Process definition Id",
    },
    Object {
      "id": "processInstanceId",
      "label": "Process instance Id",
      "title": "Process instance Id",
    },
    Object {
      "id": "variable:var1",
      "label": <React.Fragment>
        <span
          className="variableExtension"
        >
          Var
          : 
        </span>
        var1
      </React.Fragment>,
      "title": "var1",
      "type": "variables",
    },
  ],
}
`;

exports[`Process table should transform data to table compatible format 1`] = `
Object {
  "body": Array [
    Array [
      "bar",
      "foo",
      "12",
      "",
      <ForwardRef(Button)
        className="ObjectViewBtn"
        link={true}
        onClick={[Function]}
      >
        View
      </ForwardRef(Button)>,
      "--",
      "2s",
    ],
    Array [
      "abc",
      "xyz",
      "",
      "true",
      "",
      "1s",
      "--",
    ],
  ],
  "head": Array [
    Object {
      "id": "processDefinitionId",
      "label": "Process definition Id",
      "title": "Process definition Id",
    },
    Object {
      "id": "processInstanceId",
      "label": "Process instance Id",
      "title": "Process instance Id",
    },
    Object {
      "id": "variable:var1",
      "label": <React.Fragment>
        <span
          className="variableExtension"
        >
          Var
          : 
        </span>
        var1
      </React.Fragment>,
      "title": "var1",
      "type": "variables",
    },
    Object {
      "id": "variable:var2",
      "label": <React.Fragment>
        <span
          className="variableExtension"
        >
          Var
          : 
        </span>
        var2
      </React.Fragment>,
      "title": "var2",
      "type": "variables",
    },
    Object {
      "id": "variable:var3",
      "label": <React.Fragment>
        <span
          className="variableExtension"
        >
          Var
          : 
        </span>
        var3
      </React.Fragment>,
      "title": "var3",
      "type": "variables",
    },
    Object {
      "id": "flowNodeDuration:dur1",
      "label": <React.Fragment>
        <span
          className="variableExtension"
        >
          Dur
          : 
        </span>
        dur1Name
      </React.Fragment>,
      "sortable": false,
      "title": "dur1Name",
      "type": "flowNodeDurations",
    },
    Object {
      "id": "flowNodeDuration:dur2",
      "label": <React.Fragment>
        <span
          className="variableExtension"
        >
          Dur
          : 
        </span>
        dur2
      </React.Fragment>,
      "sortable": false,
      "title": "dur2",
      "type": "flowNodeDurations",
    },
  ],
}
`;
